swarm/src/lib.rs

```diff
@@ -1217,6 +1217,25 @@ where
             return Poll::Pending;
         }
     }
+
+    fn convert_port_mode(&self, port_mode: PortMode) -> TransportPortMode {
+        match port_mode {
+            PortMode::Auto => if self.listened_addrs.is_empty() {
+                TransportPortMode::New
+            } else {
+                TransportPortMode::Reuse
+            },
```

Comment:
I think we will have to make a distinction by transport here. We could grab use https://docs.rs/multiaddr/latest/multiaddr/struct.Multiaddr.html#method.protocol_stack and sanitize it by removing bits that aren't important for port-reuse like whether we are IP4 / IP6 / DNS: Essentially the first protocol.

For example, imagine we have a QUIC listen address and want to establish a TCP connection. This will fall into the `else` branch, instruct the TCP transport to reuse a port that it doesn't have (yet) and fail.

Reply:
So we first check if there is a listener for the protocol we want to dial?